librz/search/hash_search.c

```diff
@@ -68,127 +71,98 @@ static ut8 *parse_digits(const char *algo, const char *expected_digits, ut32 dig
-	if (rz_str_startswith(data->algo, "entropy")) {
-		return hsize == sizeof(double) && rz_read_be_double(calculated_hash) >= rz_read_be_double(data->digits);
-	}
-	// All the others just compare digits in with memory compare.
-	return rz_mem_eq(data->digits, calculated_hash, hsize);
+
+	// All the others just compare digest in with memory compare.
+	return rz_mem_eq(data->digest, calculated_hash, hsize);
 }
 
-static RzSearchHit *hash_and_compare(RzHashCfg *cfg, const RzSearchHashFindData *data, ut64 address, const ut8 *buffer, size_t buf_size) {
-	rz_return_val_if_fail(cfg && data && buffer, NULL);
-	if (!rz_hash_cfg_init(cfg)) {
-		RZ_LOG_ERROR("Hash config init failed.\n");
+static RzSearchHit *calculate_hash_and_compare(const SearchHashContext *data, ut64 address, const ut8 *buffer, size_t buf_size) {
+	rz_return_val_if_fail(data && data->config && buffer, NULL);
+	if (!rz_hash_cfg_init(data->config)) {
+		RZ_LOG_ERROR("search: hash config init failed.\n");
 		return NULL;
 	}
-	if (!rz_hash_cfg_update(cfg, buffer, buf_size)) {
-		RZ_LOG_ERROR("Hash config update failed.\n");
+	if (!rz_hash_cfg_update(data->config, buffer, buf_size)) {
+		RZ_LOG_ERROR("search: hash config update failed.\n");
 		return NULL;
 	}
-	if (!rz_hash_cfg_final(cfg)) {
-		RZ_LOG_ERROR("Hash config final failed.\n");
+	if (!rz_hash_cfg_final(data->config)) {
+		RZ_LOG_ERROR("search: hash config final failed.\n");
 		return NULL;
 	}
+
 	RzHashSize hsize;
-	const ut8 *calculated_hash = rz_hash_cfg_get_result(cfg, data->algo, &hsize);
-	if (hsize != data->digits_len) {
+	const ut8 *calculated_hash = rz_hash_cfg_get_result(data->config, data->algo, &hsize);
+	if (hsize != data->digest_len) {
 		rz_warn_if_reached();
 		return NULL;
 	}
+
 	if (!hashes_match(data, calculated_hash, hsize)) {
 		return NULL;
 	}
-	RzSearchHit *hit = rz_search_hit_new(data->algo, address, buf_size, NULL);
-	if (!hit) {
-		return NULL;
-	}
-	if (!rz_str_startswith(data->algo, "entropy")) {
-		// No need to add the details for those. Because the user already passed the hash value
-		// as argument. So it is known.
-		return hit;
-	}
-	RzSearchiHitDetailHash *hd = RZ_NEW(RzSearchiHitDetailHash);
-
-	hd->hash = RZ_NEWS(ut8, hsize);
-	rz_mem_copy(hd->hash, hsize, calculated_hash, hsize);
-	hd->hash_size = hsize;
-	hd->algo = rz_str_dup(data->algo);
-	hd->hash_str = rz_hash_cfg_get_result_string(cfg, data->algo, NULL, false);
-	if (!rz_search_hit_add_details(hit, RZ_SEARCH_HIT_DETAIL_HASH, hd)) {
-		RZ_LOG_ERROR("Failed to add search it details. Omit hit.\n");
-		rz_search_hit_free(hit);
-		return NULL;
-	}
-	return hit;
+
+	return rz_search_hit_new(data->algo, address, buf_size, NULL);
 }
 
 static bool hash_find(RzSearchFindOpt *fopts, void *user, ut64 address, const RzBuffer *buffer,
 	RZ_OUT RzThreadQueue *hits, RZ_OUT size_t *n_hits) {
-	RzPVector /*<RzSearchHashFindData *>*/ *pvec = (RzPVector *)user;
-
-	RzHashCfg *hcfg = init_hash_config(pvec);
```

Comment:
Just for my understanding. The `RzHashCfg` is updated with each `rz_hash_cfg_init()`, `rz_hash_cfg_update()` and `rz_hash_cfg_final()`, correct?

But here it is a shared resource and the different threads add their data to the same `RzHashCfg` instance. So don't we get race conditions?

Reply:
Hmm you might be right and this could be a race condition